dialogue-core/src/main/java/com/palantir/dialogue/core/NodeSelectionStrategyChannel.java

```diff
@@ -92,9 +94,18 @@ static LimitedChannel create(Config cf, ImmutableList<LimitedChannel> channels)
     }
 
     @Override
-    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
-        Optional<ListenableFuture<Response>> maybe = delegate.maybeExecute(endpoint, request);
-        if (!maybe.isPresent()) {
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request, boolean force) {
+        HostId hostId = request.attachments().getOrDefault(RoutingAttachments.EXECUTE_ON_HOST_ID_KEY, null);
+
+        // TODO(12345): This is broken for stats gathering, make sure to propagate this to strategies.
+        final Optional<ListenableFuture<Response>> maybe;
+        if (hostId != null) {
+            maybe = channels.get(hostId.value()).maybeExecute(endpoint, request, force);
```

Comment:
I wonder if it would be safer/easier to attach the `NodeSelectionStrategyLimitedChannel` instance instead of a HostID? Worst case we replace the client in a rebuild, but it would be impossible to accidentally switch nodes.

Reply:
See the discussion above:

> I think you suggested we could just directly send a reference to the per host channel? I think that suffers from the problem that those will be re-created when the set of URLs changes (I think we recreate all of channels then?), so even if some URLs stayed the same, the in-flight transactions would just do something weird (possible send requests to stale channels?)

I think this is the crux of the issue that is worth discussing: on a configuration change (say you add a host), you'll end up keeping a reference to the old channel, while the particular sticky channel is alive. I think your suggestion is better in terms of not screwing up the routing ever, and I don't see many downsides (we'll have a bit of a split brain for a little bit, maybe that's not the end of the world? The only problem is that these sticky channels can potentially be quite long lived (current most prevalent client can hold onto one for upwards of 5 minutes).

Curious of your thoughts

Reply:
Also second reason why I originally (on the previous PR) went with indexes: we actually have a user of #perHost channels, and I'd really like to move it over to this architecture as well (although I haven't tackled that in this PR yet), because the same user that uses sticky channels, also uses perHost for some heartbeating that needs to address nodes in the cluster.